File: Unity/Evolution2/Assets/Scripts/Net/Neuron.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Neuron : MonoBehaviour
{
    public List<float> weights =  new List<float>();
    public float activation = 0;
    public float bias = 0;
    public int weightsLength = 0;
    public List<float> deltaWeights = new List<float>();
    public float deltaBias = 0.5f;


    public void connect(List<Neuron> previousLayer){
        //float rand = 0f;
        weightsLength = previousLayer.Count;
        for(int i = 0; i < weightsLength; i++){
            weights.Add(Random.Range(-1.0f, 1.0f));
            deltaWeights.Add(Random.Range(-1.0f, 1.0f));
            //Debug.Log(weights);
        }
        bias = Random.Range(-1.0f, 1.0f);  
    }

    public void calculate(List<Neuron> previousLayer){
        float sum = 0;
        for(int i = 0; i < weightsLength; i++){
            sum += previousLayer[i].getActivation() * weights[i];
        }
        activation = function(sum + bias);
    }

    public float function(float value){
        return 1/(1 + Mathf.Exp(-value));
    }

    public float getActivation(){
        return activation;
    }

    public void setActivation(float input){
        activation = function(input);
    }

    public float getWeight(int index){
        return weights[index];
    }

    public void updateWeights(List<Neuron> previousLayer, float gradient, int x){

        float grad = 0;
        float oldDelta = 0;

        for(int i = 0; i < weightsLength; i++){
            oldDelta = deltaWeights[i];

            grad = ((oldDelta * 1.0f) + (Random.Range(-1.0f, 1.0f) * previousLayer[i].getActivation())) * gradient;
            
            deltaWeights[i] = grad;
            weights[i] += deltaWeights[i]; //+ x*0.1f;;
        }

        oldDelta = deltaBias;
        grad = ((oldDelta * 1.0f) + (Random.Range(-1.0f, 1.0f))) * gradient;
        deltaBias = grad;
        bias += deltaBias; //+ x*0.1f;

    }

    public void setDeltaWeight(int index, float input){
        deltaWeights[index] += input;
    }

    public void setWeight(int index, float input){
        weights[index] += input;
    }

    public void setDeltaBias(float input){
        deltaBias += input;
    }

    public void setBias(float input){
        bias += input;
    }
}

File: C++/DBZFighter/DBZFighter/neuron.cpp
#include "neuron.h"

float sigmoid(float input) {
	return 1 / (1 + exp(-input));
}
neuron::neuron() {
	bias = 0;
	activation = 0;
	weightLength = 0;
	outputted = false;
}
neuron::~neuron() {

}
void neuron::setWeight(int index, float value) {
	weights[index] = value;
}
float neuron::getWeight(int index)
{
	return weights[index];
}
void neuron::addWeight(float value) {
	weights.push_back(value);
	weightLength += 1;
}
void neuron::calculate() {
	float sum = 0.f;
	for (int i = 0; i < weightLength; i++) {
		sum += weights[i] * inputs[i]->getActivation();
	}
	activation = sigmoid(sum + bias);
}
float neuron::getActivation() {
	return activation;
}
void neuron::setActivation(float act) {
	activation = act;
}
void neuron::addInput(neuron* node) {
	inputs.push_back(node);
	node->outSet();
}

void neuron::outSet() {
	outputted = true;
}